Use numeric_cast<mp_integer> instead of deprecated to_integer variants [blocks: #3800]

tautschnig opened the pull request:
to_integer should only be used with a constant_exprt as first argument. Includes
a bugfix in goto_rw where the result of pointer_offset_bits did not have
operator* applied.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use numeric_cast<mp_integer> instead of deprecated to_integer variants

to_integer should only be used with a constant_exprt as first argument. Includes
a bugfix in goto_rw, where the result of pointer_offset_bits did not have
operator* applied.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 51112a8525a0cded93e60049d33935fc4e24d49d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98806196

martin-cs reviewed (approved):
Looks fine from my point of view.

thk123 commented on regression/cbmc-library/memmove-01/test.desc:
```diff
@@ -1,8 +1,13 @@
-CORE
+KNOWNBUG
```
:pick: The commit message summary does not suggest a test is being disbaled - could this be moved into a separate commit.

thk123 commented on src/analyses/goto_rw.cpp:
```diff
@@ -150,22 +150,21 @@ void rw_range_sett::get_objects_byte_extract(
-    index*=8;
-    if(index>=pointer_offset_bits(be.op().type(), ns))
+    *index *= 8;
+    if(*index >= *pointer_offset_bits(be.op().type(), ns))
```
How did this even compile? Was it being implictly converted? In which case, do we only have to explicitly convert since they are now the same type?

kroening replied:
That's because of the overloaded >= operator. I don't like it, but it's part of C++17.

thk123 replied:
Could you elaborate, I don't really understand? `pointer_offset_bits` returns an `optionalt<mp_integer>`

tautschnig replied:
`optionalt` has the following:
````c++
template< typename T > bool operator>=( optional<T> const &, nullopt_t )
{
    return true;
}

template< typename T > bool operator>=( nullopt_t, optional<T> const & x )
{
    return (!x);
}
````
So you can compare to about anything...

thk123 replied:
lol what possible reason could there be for those operators?!

tautschnig replied:
No idea - in particular http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3848.html#optional does not seem to mention those.

thk123 replied:
Wait but hang on, in the original, index is an `mp_integer`, which I suppose can be implicitly converted to an `optionalt<mp_integer>`, but surely the `optionalt` can't be implicitly converted to the `nullopt_t`?

Also, on that link, aren't these exactly [2.9](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3848.html#optional.nullops) on the linked ref

tautschnig replied:
I'm sorry, I was wrong, it's actually 2.10 from the reference that we need to look at, and currently for us that's:
````c++
template< typename T > bool operator<( optional<T> const & x, optional<T> const & y )
{
    return (!y) ? false : (!x) ? true : *x < *y;
}
````
and
````c++
template< typename T > bool operator>=( optional<T> const & x, optional<T> const & y )
{
    return !(x < y);
}
````
that matter here. So the comparison was silently `false` in a case that we should actually fail on.

thk123 commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -566,8 +566,7 @@ void value_sett::get_value_set_rec(
     object_mapt pointer_expr_set;
-    mp_integer i;
-    bool i_is_set=false;
+    optionalt<mp_integer> i;
```
:green_heart:

thk123 commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -233,16 +233,16 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
 
     auto element_width = pointer_offset_bits(subtype, ns);
-    mp_integer num_elements;
+    const auto num_elements = numeric_cast<mp_integer>(array_type.size());
```
I can't spot the bug you've fixed here that causes the test to fail - could you pull it out into a standa lone commit for ease of reviewing and when that test envitably gets `git blame`d to see why it is turned off.

tautschnig replied:
Done, it's now the first (very small) commit.

thk123 reviewed (commented):
:pick: When doing (un)ordered lists in commit messages, please add a new line for ease of reading:
```
to_integer should only be used with a constant_exprt as first argument. Includes
two bugfixes: 
1) in goto_rw, where the result of pointer_offset_bits did not
have operator* applied
2) in lower_byte_extract, where the result of
to_integer was not handled correctly; the latter, however, also means that we
cannot solve memmove-01 at the moment.
```

Please move the test disabling and corresponding bug fix to a separate commit. Also, how come the test passed if the behaviour was wrong? Can it be tighted up so that it wouldn't pass before.

tautschnig commented:
> Also, how come the test passed if the behaviour was wrong? Can it be tighted up so that it wouldn't pass before.

Kind of an obvious question that I should have asked myself... I'll do as you suggested and will create a separate commit.

tautschnig committed:
When to_integer returns true, the result must not be used

The mp_integer will be zero, which meant that the loop terminated immediately,
and thus the array was not populated. This only affected parts of the array
being copied around, which thus had non-deterministic values. Extending the
regression test to compare all values confirmed this problem.

thk123 commented:
Thanks!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cf1f65ae4e89d97f319e62c69283011e59aae0ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99163020
